File: AoC_Scala3/src/day15/day15.scala
package day15

import AoC_Lib.*

object day15 {
  val grid: Map[Pos, Int] = inputStrs(day = 15)
    .zipWithIndex
    .flatMap((s,y) => s
      .zipWithIndex
      .map((c,x) => Pos(x,y) -> c.asDigit))
    .toMap

  def dijkstraSearch(grid: Map[Pos,Int], start: Pos, target: Pos): Option[Int] = {
    import scala.collection.*
    val visited: mutable.Map[Pos, Int] = mutable.Map.empty
    val toVisit: mutable.PriorityQueue[(Pos, Int)] = mutable.PriorityQueue.empty(Ordering.by(-_._2))

    toVisit.enqueue(start -> 0)
    while (toVisit.nonEmpty) {
      val (node, dist) = toVisit.dequeue()
      if (!visited.contains(node)) {
        visited(node) = dist
        if (node == target) return Some(dist)
        node.near4
          .filterNot(visited.contains)
          .filterNot(grid(_) <= 0)
          .foreach(p => toVisit.enqueue(p -> (dist + grid(p))))
      }
    }
    None
  }

  val maxPosInGrid: Pos = grid.keySet.reduce(_ max _)

  def part1: Int = dijkstraSearch(grid.withDefaultValue(-1),Pos.zero,maxPosInGrid).get

  val fullgrid: Map[Pos, Int] = grid.withDefault{
    p => if p.x < 0 || p.y < 0 || p.x > (maxPosInGrid.x + 1) * 5 - 1 || p.y > (maxPosInGrid.y + 1) * 5 - 1 then -1
    else (grid(Pos(p.x % (maxPosInGrid.x + 1), p.y % (maxPosInGrid.y + 1))) + p.x / (maxPosInGrid.x + 1) + p.y / (maxPosInGrid.y + 1) - 1) % 9 + 1
  }

  def part2: Long = dijkstraSearch(fullgrid,Pos.zero,5*:(maxPosInGrid+Pos(1,1))-Pos(1,1)).get

  @main
  def Chiton(): Unit = {
    Console.println(part1)     // 656
    Console.println(part2)     // 2979
  }

}

File: Aoc_Kotlin/src/main/kotlin/day15/findPath/Day15.kt
package day15.findPath

import AoC_Lib.*
import java.util.*

object Day15: SomeDay(2021,15) {
    override val title = "Chiton"

    private fun dijkstraSearch(grid: ArrayGrid<Int>, start: Pos, target: Pos): Int? {
        val visited = mutableMapOf<Pos,Int>()
        val toVisit = PriorityQueue<Pair<Pos,Int>> { o1, o2 -> o1.second - o2.second }

        toVisit.add(start to 0)
        while (toVisit.isNotEmpty()) {
            val (node, dist) = toVisit.poll()
            if (!visited.contains(node)) {
                visited[node] = dist
                if (node == target) return dist
                node.near(NearDir.Axis)
                    .filterNot{ visited.contains(it) }
                    .filter{ grid.contains(it) }
                    .forEach{ p -> toVisit.add(p to (dist + grid[p.x][p.y]))}
            }
        }
        return null
    }

    private fun dpSearch(grid: ArrayGrid<Int>, start: Pos, end: Pos): Int {
        val pq = PriorityQueue<Pair<Pos,Int>> { a, b -> a.second - b.second }
        val risk = Array(grid.size) { Array(grid.size) { Int.MAX_VALUE } }
        risk[0][0] = 0
        pq.add(start to 0)
        while (pq.isNotEmpty()) {
            val (current, dist) = pq.poll()
            current.near(NearDir.Axis)
                .forEach { p ->
                    if (grid.contains(p) && risk[p.x][p.y] > dist + grid[p.x][p.y]) {
                        risk[p.x][p.y] = dist + grid[p.x][p.y]
                        pq.add(p to risk[p.x][p.y])
                    }
                }
        }
        return risk[end.x][end.y]
    }

    override fun first(data: String): Any? {
        val grid = data.lines()
            .map { it.toCharArray()
                .map { it.digitToInt() }
                .toTypedArray() }
            .toTypedArray()
        return dijkstraSearch(grid,Pos.Zero,Pos(grid.lastIndex,grid.last().lastIndex))
    } // 656 Time: 101ms

    override fun second(data: String): Any? {
        val scale = 5
        val grid = data.lines()
            .map { it.toCharArray()
                .map { it.digitToInt() }
                .toTypedArray() }
            .toTypedArray()
        val fullGrid = (0 until grid.size * scale).map { i ->
            (0 until grid.size * scale).map { j ->
                (grid[i % grid.size][j % grid.size] + (i / grid.size + j / grid.size))
                    .let { if (it < 10) it else it - 9 }
            }.toTypedArray()
        }.toTypedArray()
        //return dijkstraSearch(fullGrid,Pos.Zero,Pos(fullGrid.lastIndex,fullGrid.last().lastIndex))
        return dpSearch(fullGrid,Pos.Zero,Pos(fullGrid.lastIndex,fullGrid.last().lastIndex))
    } // 2979 Time: 1026ms
      // 2979 Time: 178ms
}

fun main() = SomeDay.mainify(Day15)